src/main/scala/is/hail/utils/richUtils/RichIndexedRowMatrix.scala

```diff
@@ -0,0 +1,160 @@
+  */
+
+object RichIndexedRowMatrix {
+
+  implicit class RichIndexedRowMatrix(indexedRowMatrix: IndexedRowMatrix) {
+    def toBlockMatrixDense(): BlockMatrix = {
+      toBlockMatrixDense(1024, 1024)
+    }
+
+    def toBlockMatrixDense(rowsPerBlock: Int, colsPerBlock: Int): BlockMatrix = {
+      require(rowsPerBlock > 0,
+        s"rowsPerBlock needs to be greater than 0. rowsPerBlock: $rowsPerBlock")
+      require(colsPerBlock > 0,
+        s"colsPerBlock needs to be greater than 0. colsPerBlock: $colsPerBlock")
+      val m = indexedRowMatrix.numRows().toInt
+      val n = indexedRowMatrix.numCols().toInt
+
+      val blocksRDD: RDD[((Int, Int), Matrix)] = indexedRowMatrix.rows.flatMap({ ir =>
+        val blockRow = ir.index / rowsPerBlock
+        val rowInBlock = ir.index % rowsPerBlock
+
+        val partitionedArray: Iterator[Array[Double]] = ir.vector.toArray.grouped(colsPerBlock)
+        val partitionedArrayWithIndices = partitionedArray.zipWithIndex
+
+        partitionedArrayWithIndices.map(tuple =>
+          tuple match {
+            case (values, blockColumn) =>
+              ((blockRow.toInt, blockColumn), (rowInBlock.toInt, values))
+          })
+      }).groupByKey(new GridPartitioner(m, n, rowsPerBlock, colsPerBlock)).map(tuple =>
+        tuple match {
+          case ((blockRow, blockColumn), itr) =>
+            val actualNumRows =
+              if (m - blockRow * rowsPerBlock < rowsPerBlock) {
+                m - blockRow * rowsPerBlock
+              } else {
+                rowsPerBlock
+              }
+            val actualNumColumns =
+              if (n - blockColumn * colsPerBlock < colsPerBlock) {
+                n - blockColumn * colsPerBlock
+              } else {
+                colsPerBlock
+              }
+
+            val arraySize = actualNumRows * actualNumColumns
+            val matrixAsArray = new Array[Double](arraySize)
+            itr.foreach(element => element match {
+              case (rowWithinBlock, values) =>
+                var i = 0
+                while (i < values.length) {
+                  matrixAsArray.update(i * actualNumRows + rowWithinBlock, values(i))
+                  i += 1
+                }
+            })
+            ((blockRow, blockColumn), new DenseMatrix(actualNumRows, actualNumColumns, matrixAsArray, false))
+        })
+      new BlockMatrix(blocksRDD, rowsPerBlock, colsPerBlock)
+    }
+  }
+}
+
+
+/**
+  * A grid partitioner, which uses a regular grid to partition coordinates.
+  *
+  * NOTE: This is ripped directly from Apache Spark's BlockMatrix class. It's private to mllib, so I copied it here.
+  *
+  * @param rows Number of rows.
+  * @param cols Number of columns.
+  * @param rowsPerPart Number of rows per partition, which may be less at the bottom edge.
+  * @param colsPerPart Number of columns per partition, which may be less at the right edge.
+  */
+private class GridPartitioner(
```

Comment:
Doing the same tests as above, multiplying takes on average 38-40 seconds. Granted, this is a trial with all of the changes suggested here implemented, so the seeming slight decrease in speed is probably just a result of significantly fewer partitions in use now that I fixed the GridPartitioner constructor (This is 60 partitions on 128 cores now, as opposed to 900 partitions on 128 cores when my GridPartitioner was constructed wrong).